File: cmpe101/Find Graph Components/List.c
/*
  Herman Chin
  hchin@ucsc.edu
  Summer 2011
  CMPS 101
  Tantalo
  Lab 5 
*/

#include <stdio.h>
#include <stdlib.h>
#include "List.h"

typedef struct Node{
	int data;
	struct Node* next;
	struct Node* prev;
} Node;
typedef Node* NodeRef;

typedef struct List{
	NodeRef curr;
	NodeRef head;
	NodeRef tail;
	int size;
} List;


/*** Constructors-Destructors ***/
NodeRef newNode(int node_data){
        NodeRef N = malloc(sizeof(Node));
        N->data = node_data;
        N->next = NULL;
        N->prev = NULL;
        return(N);
}

void freeNode(NodeRef* pN){
        if( pN!=NULL && *pN!=NULL ){
                free(*pN);
                *pN = NULL;
        }
}

ListRef newList(void){
	ListRef L;
	L = malloc(sizeof(List));
	L->curr = NULL;
	L->head = NULL;
	L->tail = NULL;
	L->size = 0;
	return(L);
}

void freeList(ListRef* pL){
	if( pL!=NULL && *pL!=NULL ){
		while(!isEmpty(*pL)){
			deleteFirst(*pL);
		}
		free(*pL);
		*pL = NULL;
	}
}
/*** Access functions ***/
int isEmpty(ListRef L){
	if(L == NULL){
		printf("List Error: isEmpty()isEmpty()");
		exit(1);
	}
	return(L->size == 0);
}
int offEnd(ListRef L){
	return(L->curr == NULL);
}
int atFirst(ListRef L){
        if(isEmpty(L)){
                printf("List error: isEmpty");
                exit(1);
        }
	return(L->curr == L->head);
}

int atLast(ListRef L){
        if(isEmpty(L)){
                printf("List error: isEmpty,atLast");
                exit(1);
        }
	return(L->curr == L->tail);
}
int getFirst(ListRef L){
        if(isEmpty(L)){
                printf("List error: isEmpty");
                exit(1);
        }
	return(L->head->data);
}
int getLast(ListRef L){
        if(isEmpty(L)){
                printf("List error: isEmpty");
                exit(1);
        }
	return(L->tail->data);
}

int getCurrent(ListRef L){
        if(isEmpty(L)){
                printf("List error: isEmpty,getCurrent");
                exit(1);
        }
	return(L->curr->data);
}
int getLength(ListRef L){
        if(isEmpty(L)){
		return(0);
                printf("List error: isEmpty,getLength");
                exit(1);
        }
	return(L->size);
}
/* ret = 1 true, ret = 0 false*/
int equals(ListRef A, ListRef B){
	int ret = 0;
	if(A->head != NULL && B->head != NULL){
		moveFirst(A);
		moveFirst(B);
		while(A->curr != A->tail && B->curr != B->tail){
			if(A->curr->data == B->curr->data){
				ret = 1;
			}else{
				ret = 0;
			}
			A->curr = A->curr->next;
			B->curr = B->curr->next;
		}
                if(A->curr->data == B->curr->data){
                        ret = 1;
                }else{
                        ret = 0;
                }
	}
	return(ret);
};

/*** Manipulation procedures ***/
void makeEmpty(ListRef L){
	if(isEmpty(L)){
		printf("List error: isEmpty error");
		exit(1);
	}
	moveFirst(L);
	while(L->curr != L->tail){
		L->curr->data = 0;		
		L->curr = L->curr->next;
	}
	L->curr = L->head = L->tail = NULL;
	L->size = 0;
}
void moveFirst(ListRef L){
        if(isEmpty(L)){
                printf("List error: isEmpty,moveFirst");
                exit(1);
        }
	L->curr = L->head;
}
void moveLast(ListRef L){
        if(isEmpty(L)){
                printf("List error: isEmpty");
                exit(1);
        }
	L->curr = L->tail;
        if(offEnd(L)){
                printf("List error: moveLast");
                exit(1);
        }
}
void movePrev(ListRef L){
        if(isEmpty(L)){
                printf("List error: isEmpty");
                exit(1);
        }
	if(L->curr->prev != NULL){
	        L->curr = L->curr->prev;
	}
        if(offEnd(L)){
                printf("List error: movePrev");
                exit(1);
        }
}
void moveNext(ListRef L){
        if(isEmpty(L)){
                printf("List error: isEmpty,moveNext");
                exit(1);
        }
        if(L->curr->next != NULL){
                L->curr = L->curr->next;
        }
        if(offEnd(L)){
                printf("List error: movePrev");
                exit(1);
        }
}
void insertBeforeFirst(ListRef L, int data){
        NodeRef newN = newNode(data);
	if(offEnd(L)){
		L->curr = L->head = L->tail = newN;
	}else{
	        moveFirst(L);
		L->curr->prev = newN;
		newN->next = L->curr;
		L->head = newN;
	}
        L->size++;
}
void insertAfterLast(ListRef L, int data){
        NodeRef newN = newNode(data);
        if(offEnd(L)){
                L->curr = L->head = L->tail = newN;
        }else{
	        L->tail->next = newN;
	        newN->prev = L->tail;
		L->tail = newN;
	}
        L->size++;
}
void insertBeforeCurrent(ListRef L, int data){
        NodeRef newN = newNode(data);
        if(offEnd(L)){
		L->curr = L->head = L->tail = newN;
        }
	if(L->curr == L->head){
		insertBeforeFirst(L,data);
	}else{
		newN->prev = L->curr->prev;
		newN->next = L->curr;
		L->curr->prev->next = newN;
		L->curr->prev = newN;
	}
        L->size++;
}
void insertAfterCurrent(ListRef L, int data){
        NodeRef newN = newNode(data);
        if(offEnd(L)){
                L->curr = L->head = L->tail = newN;
        }
        if(L->curr == L->tail){
                insertAfterLast(L,data);
        }else{
                newN->next = L->curr->next;
                newN->prev = L->curr;
                L->curr->next->prev = newN;
                L->curr->next = newN;
        }
	L->size++;
}
void deleteFirst(ListRef L){
	NodeRef N = L->head;
	if(N->next == NULL ){
		L->head = L->curr = L->tail = NULL;
	}else{
		L->head = L->head->next;
		L->head->prev = NULL;
	}
	L->size--;
        freeNode(&N);
}
void deleteLast(ListRef L){
        NodeRef N = L->tail;
        if(N->prev == NULL){
		L->tail = NULL;
	}else{
		L->tail = L->tail->prev;
		L->tail->next = NULL;
        }
        L->size--;
	freeNode(&N);
}
void deleteCurrent(ListRef L){
        if(isEmpty(L) || offEnd(L)){
                printf("List error: isEmpty,deleteCurrent");
                exit(1);
        }
	if(L->curr == L->head){
		deleteFirst(L);
	}else if(L->curr == L->tail){
		deleteLast(L);
	}else{
        	NodeRef N = L->curr;
		L->curr->prev->next = L->curr->next;
		L->curr->next->prev = L->curr->prev;
        	L->size--;
        	freeNode(&N);
		L->curr = NULL;
	}
}
/*** Other operations ***/
void printList(FILE* out, ListRef L){
	if(isEmpty(L) || offEnd(L)){
                printf("List error: printList");
                exit(1);
        }
	NodeRef p = L->head;
	while(p->next != NULL){
		fprintf(out,"%d ",p->data);
		p = p->next;
	}
        fprintf(out,"%d ",p->data);
	fprintf(out,"\n");
}
ListRef copyList(ListRef L){
	moveFirst(L);
	ListRef N;
	N = malloc(sizeof(List));
	N->curr->data = L->curr->data;
	while(L->curr != NULL){
		insertAfterCurrent(N,L->curr->data);
		L->curr = L->curr->next;
		N->curr = N->curr->next;
	}
	return(N);
}

File: cmpe101/Matrix_Graph_Shortest_Path/List.java
/*
  Herman Chin
  hchin
  Program 3 
  CS101 Summer 2011
  Tantalo
*/

public class List{

	public class Node{
		Object data;
		Node next;
		Node prev;
		Node(Object data){this.data = data; next = null; prev = null;}
		public String toString(){return String.valueOf(data);}
	}
// Fields
	public Node head;
	public Node tail;
	public Node current;
	public int length;
// Constructor
	List(){head = tail = current = null;length = 0;} 

        public String toString(){
           String str = "";
           for(Node N=head; N!=null; N=N.next){
              str += N.toString() + " ";
           }
           return str;
        }

// Access Fucntions

	boolean isEmpty(){
		return length == 0;
	}
	boolean offEnd(){
		if(current == null) return true;
		return false; 
	}
	boolean atFirst(){
		if(head == null) return true;
		return false;
	}
	boolean atLast(){
		if(tail == null) return true;
		return false;
	}
        Object getFirst(){
		if(isEmpty()){
			throw new RuntimeException("Error: getFirst");
		}
		return head.data;
	}
	Object getLast(){
                if(isEmpty()){
                        throw new RuntimeException("Error: getLast");
                }
                return tail.data;
	}
	Object getCurrent(){
                if(isEmpty()){
                        throw new RuntimeException("Error: getCurrent");
                }
                if(offEnd()){
                        throw new RuntimeException("Error: getCurrent");
                }
		return current.data;
	}
	Object getLength(){
		return length;
	}
	boolean equals(List x){
		Node A = head;
		Node B = x.head;
		while(A!=null && B!=null){
			if(A.data != B.data){
				return false;
			}
			A = A.next;
			B = B.next;
		}
		return true;
	}
	void makeEmpty(){
		head = tail = current = null;
		length = 0;
                if(isEmpty()){
                       System.out.print("Emptied");
                }
	}	
	
	void moveFirst(){
                if(isEmpty()){
                        throw new RuntimeException("Error: moveFirst");
                }
		current = head;
                if(offEnd()){
                        throw new RuntimeException("Error: offend moveFirst");
                }
	}
	void moveLast(){
                if(isEmpty()){
                        throw new RuntimeException("Error: moveLast");
                }
		current = tail;
                if(offEnd()){
                        throw new RuntimeException("Error: moveLast");
                }
	}
	void movePrev(){
		if(isEmpty()){
                        throw new RuntimeException("Error: movePrev");
                }
		current = current.prev;
                if(offEnd()){
                        throw new RuntimeException("Error: movePrev");
                }
	}
	void moveNext(){
                if(isEmpty()){
                        throw new RuntimeException("Error: moveNext");
                }
                current = current.next;
                if(offEnd()){
                        throw new RuntimeException("Error: moveNext");
                }
	}
	void insertBeforeFirst(Object data){
                Node newNode = new Node(data);
		if(!isEmpty()){
			newNode.next = head;
			head.prev = newNode;
			head = newNode;
		}else{
			tail = head = current = newNode;
		}
		length++;
                if(offEnd()){
                        throw new RuntimeException("Error: insertBeforeFirst");
                }
	}
	void insertAfterLast(Object data){
                Node newNode = new Node(data);
		if(!isEmpty()){
                	newNode.prev = tail;
			tail.next = newNode;
			tail = newNode;
		}else{
			head = tail = current = newNode;
		}
		length++;
                if(offEnd()){
                        throw new RuntimeException("Error: insertAfterLast");
                }
	}
	void insertBeforeCurrent(Object data){
                Node newNode = new Node(data);
		if(!isEmpty() && current != head){
			newNode.prev = current.prev;
			current.prev.next = newNode;
			newNode.next = current;
			current.prev = newNode;
		}else if(current == head){
			insertBeforeFirst(data);
		}else{
			head = tail = current = newNode;
		}
		length++;
                if(offEnd()){
                        throw new RuntimeException("Error: insertBeforeCurrent");
                }
	}
	void insertAfterCurrent(Object data){
                Node newNode = new Node(data);
		if(!isEmpty() && current != tail){
	       	        newNode.next = current.next;
	       	        current.next.prev = newNode;
	                newNode.prev = current;
	                current.next = newNode;
		}else if(current == tail){
			insertAfterLast(data);
                }else{
                        head = tail = current = newNode;
                }
		length++;
                if(offEnd()){
                        throw new RuntimeException("Error: insertAfterCurrent");
                }
	}
	void deleteFirst(){
                if(isEmpty()){
                        throw new RuntimeException("Error: deleteFirst");
                }
		head = head.next;
		head.prev = null;
		length--;
	}
	void deleteLast(){
                if(isEmpty()){
                        throw new RuntimeException("Error: deleteLast");
                }
		tail = tail.prev;
		tail.next = null;
		length--;
	}
	void deleteCurrent(){
                if(isEmpty()){
                        throw new RuntimeException("Error: deleteCurrent");
                }
                if(offEnd()){
                        throw new RuntimeException("Error: deleteCurrent");
                }
		if(current != head && current != tail){
			current.next.prev = current.prev;
			current.prev.next = current.next;
		}else if(current == head){
			deleteFirst();
		}else if(current == tail){
			deleteLast();
		}
		current = current.next;
		length--;
                if(offEnd()){
                        throw new RuntimeException("Error: deleteCurrent");
                }
	}
}
